wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.Beta;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.Page;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.devfile.Devfile;
+import org.eclipse.che.api.core.model.workspace.devfile.UserDevfile;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.devfile.server.model.impl.UserDevfileImpl;
+import org.eclipse.che.api.devfile.server.spi.UserDevfileDao;
+import org.eclipse.che.api.devfile.shared.event.DevfileCreatedEvent;
+import org.eclipse.che.api.devfile.shared.event.DevfileDeletedEvent;
+import org.eclipse.che.api.devfile.shared.event.DevfileUpdatedEvent;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.lang.NameGenerator;
+import org.eclipse.che.commons.lang.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Facade for {@link UserDevfile} related operations. */
+@Beta
+@Singleton
+public class UserDevfileManager {
+  private static final Logger LOG = LoggerFactory.getLogger(UserDevfileManager.class);
+  private final UserDevfileDao userDevfileDao;
+  private final EventService eventService;
+
+  @Inject
+  public UserDevfileManager(UserDevfileDao userDevfileDao, EventService eventService) {
+    this.userDevfileDao = userDevfileDao;
+    this.eventService = eventService;
+  }
+
+  /**
+   * Stores {@link Devfile} instance
+   *
+   * @param devfile instance of devfile which would be stored
+   * @return new persisted devfile instance
+   * @throws ConflictException when any conflict occurs (e.g Devfile with such name already exists
+   *     for {@code owner})
+   * @throws ServerException when any other error occurs
+   */
+  public UserDevfile createDevfile(Devfile devfile)
+      throws ServerException, NotFoundException, ConflictException {
+    requireNonNull(devfile, "Required non-null devfile");
+    UserDevfile result =
+        userDevfileDao.create(
+            new UserDevfileImpl(NameGenerator.generate("userdevfile", 16), devfile));
```

Comment:
it's the id visible for the user, right? As we agreed let's avoid exposing userdevfile to a user. Let's use just `devfile` prefix